getting started and tutorial - establishing consensus (#27)
onmax/nimiq-developer-center: build/web-client/tutorials/establishing-consensus.md

## Overview

- Nimiq is a censorship-resistant, privacy-oriented, and decentralized blockchain-based payment protocol
- A blockchain is a shared digital ledger that is kept up-to-date across all users in the network through a process called establishing consensus
- Compared to most other blockchain systems, it allows a direct connection to the network using a light node
- These light nodes use Zero Knowledge Proofs (ZKPs) to efficiently verify blockchain transactions and states

## Learning

Nimiq is a blockchain-based payment protocol. This technology offers several key benefits:

- **Censorship-resistant:** No one can prevent you from using it
- **Privacy-oriented:** You don’t need to provide any personal information to open an account
- **Decentralized:** It’s not controlled by any single person but by all its users

A blockchain is like a shared digital ledger. Keeping this ledger up to date across all users in the network is called establishing consensus. Traditionally, most blockchain systems require you to connect through a third party to avoid downloading the entire blockchain, which can be over 100 GB in size. This approach necessitates trusting the third party.

> Nimiq, however, allows direct connection to the network using a light node. Unlike full nodes, which download and verify the entire blockchain, light nodes only download a small portion of it. They rely on full nodes to provide the necessary information to verify transactions. This approach significantly reduces storage and computational requirements, making light nodes ideal for devices with limited resources, such as browsers or mobile devices.

Reviewer: ```suggestion
Nimiq allows direct connection to the network with Zero Knowledge Proofs (ZKPs) for both light and full nodes. Unlike history nodes, which download and verify the entire blockchain, light and full nodes only download the necessary data to verify the [macro chain's](/learn/protocol/ZKP-and-recursive-SNARKs.md) validity. This approach significantly reduces storage and computational requirements, making light nodes ideal for devices with limited resources, such as browsers or mobile devices, and making full nodes much lighter.
```

> Nimiq, however, allows direct connection to the network using a light node. Unlike full nodes, which download and verify the entire blockchain, light nodes only download a small portion of it. They rely on full nodes to provide the necessary information to verify transactions. This approach significantly reduces storage and computational requirements, making light nodes ideal for devices with limited resources, such as browsers or mobile devices.

Reviewer: Although this guide is for a web client, which is a light node, both light and full nodes sync with ZKPs. I think it's worth mentioning that before continuing exclusively with the light node reference. Wdyt?

> In Nimiq, these light nodes use Zero Knowledge Proofs (ZKPs) to efficiently verify blockchain transactions and states, providing a secure and user-friendly experience.

Reviewer: Because above the paragraph is quite complete, I would remove this sentence

<div dark:bg-gray-300 bg-gray-100 p-16 my-48 rounded-6>
    <img class="object-contain max-h-[max(80vh,220px)]" src="/assets/images/tutorials/nimiq-network.svg" alt="Illustration of the nimiq network compared to most other blockchains" />
</div>

### Using `@nimiq/core` to Establish Consensus

You have the option to interact with the Nimiq blockchain using either a web browser or Node.js, through the `@nimiq/core` npm module. Begin by setting up your project as usual, then install the module using npm.

```bash
npm install @nimiq/core@next
```

In order to connect to the network and establish consensus, you need to import the Nimiq Web package and configure the client.

```js
const Nimiq = require('@nimiq/core')
```

- To create a client configuration, use the `ClientConfiguration()` constructor
- To configure the client to use the Nimiq Testnet use `config.network('TestAlbatross')`
- To create an instance of the Client call `Client.create(config.build())`

```js
const config = new Nimiq.ClientConfiguration() // Create a configuration builder
config.network('TestAlbatross') // Select Testnet

> const client = await Nimiq.Client.create(config.build()) // Instantiate and launch the client

Reviewer 2: What about putting comments on top to avoid scrollbar-x?

```suggestion
// Create a configuration builder
const config = new Nimiq.ClientConfiguration()

// Select Testnet
config.network('TestAlbatross')

// Instantiate and launch the client
const client = await Nimiq.Client.create(config.build())
```

```

> These three lines are all you need to connect to the network and establish consensus. All you have to do is wait until the consensus is established. You can check the consensus state by calling `isConsensusEstablished()` which returns `true` if consesus is established.

Reviewer 2: ```suggestion
These three lines are all you need to connect to the network and establish consensus. All you have to do is wait until the consensus is established. By calling `isConsensusEstablished()`, which returns a boolean, you can check whether you have established consensus or not.
```

Something like this or similar. Instead of using "consensus state"

```js

> const consensus = await client.isConsensusEstablished()

Reviewer 2: ```suggestion
const hasConsensus = await client.isConsensusEstablished()
```

`consensus` itself is a state like `pending`, `syncing` or `established`. If we add the "has" is more clear is boolean, no?

```

Great! You just learned how to configure a client and establish consensus. Now let's practice what you just learned!

## Practice

Create a new directory and the `@nimiq/core` package:

```

> mkdir nimiq-2-0-tutorial

Reviewer 2: what about just `nimiq-tutorial`?

cd nimiq-2-0-tutorial
npm init -y
npm install @nimiq/core@next
```

Create a new file `index.js` and configure the client:

```js

> const Nimiq = require('@nimiq/core')

Reviewer 2: I prefer using esm over cjs.

I would use `import {} from ''`
